eliot/_action.py

```diff
@@ -415,6 +427,293 @@ def __exit__(self, type, exception, traceback):
+            C{FAILED_STATUS}.
+        @raise InvalidStartMessage: If C{start_message} does not have a
+            C{ACTION_STATUS_FIELD} of C{STARTED_STATUS}, or if it has a
+            C{task_level} indicating that it is not the first message of an
+            action.
+
+        @return: A new C{WrittenAction}.
+        """
+        if start_message.contents.get(ACTION_STATUS_FIELD, None) != STARTED_STATUS:
+            raise InvalidStartMessage.wrong_status(start_message)
+        if start_message.task_level.level[-1] != 1:
+            raise InvalidStartMessage.wrong_task_level(start_message)
+        action = cls(
+            start_message=start_message,
+            _children=pmap(),
+            end_message=None,
+        )
+        for child in children:
+            action = action._add_child(child)
+        if end_message:
+            return action._end(end_message)
+        return action
+
+    @property
+    def action_type(self):
+        """
+        The type of this action, e.g. C{"yourapp:subsystem:dosomething"}.
+        """
+        return self.start_message.contents[ACTION_TYPE_FIELD]
+
+    @property
+    def status(self):
+        """
+        One of C{STARTED_STATUS}, C{SUCCEEDED_STATUS}, or C{FAILED_STATUS}.
+        """
+        message = self.end_message if self.end_message else self.start_message
+        return message.contents[ACTION_STATUS_FIELD]
+
+    @property
+    def task_uuid(self):
+        """
+        The UUID of the task to which this action belongs.
+        """
+        return self.start_message.task_uuid
+
+    @property
+    def task_level(self):
+        """
+        The level of the task in which the action occurs.
+        """
+        return self.start_message.task_level.parent()
+
+    @property
+    def start_time(self):
+        """
+        The Unix timestamp of when the action started.
+        """
+        return self.start_message.timestamp
+
+    @property
+    def end_time(self):
+        """
+        The Unix timestamp of when the action ended, or C{None} if there has been
+        no end message.
+        """
+        if self.end_message:
+            return self.end_message.timestamp
+
+    @property
+    def exception(self):
+        """
+        If the action failed, the name of the exception that was raised to cause
+        it to fail. If the action succeeded, or hasn't finished yet, then
+        C{None}.
+        """
+        if self.end_message:
+            return self.end_message.contents.get(EXCEPTION_FIELD, None)
+
+    @property
+    def reason(self):
+        """
+        The reason the action failed. If the action succeeded, or hasn't finished
+        yet, then C{None}.
+        """
+        if self.end_message:
+            return self.end_message.contents.get(REASON_FIELD, None)
+
+    @property
+    def children(self):
+        """
+        The list of child messages and actions, excluding the start and end
```

Comment:
Children have a particular order, so I suggest sorting by task_level (task_level of messages for messages, task level of start message for actions).